### PR TITLE
I adjusted the simulation process towards state specific transition probabilities.

### PR DESCRIPTION
**Pull request by MaxBlesch**
The function requires now EV. The agents expectation on the future value of the bus. Furthermore it requires the real transition matrix governing the process.

### CODE REVIEW

**Comment by peisenha** [comment]
I do not review PR with failing CI.

**Comment by MaxBlesch** [comment]
The code passed on my machine and the error was due to the lower memory of travis. I now extracted the loop over buses from the numba loop, such that I draw the increments for each bus separate. Thus the increments array is now shaped (num_states, num_periods). We could discuss this Tuesday and maybe you know a better solution.

File: ruspy/estimation/estimation.py [hunk 1]
```diff
@@ -42,6 +42,10 @@ def estimate(init_dict, df, maint_func=lin_cost, repl_4=False):
     states = df.loc[:, "state"].to_numpy()
     num_obs = df.shape[0]
     num_states = init_dict["states"]
+    if init_dict["maint_func"] == "linear":
```
Comment (peisenha) [inline ruspy/estimation/estimation.py]: Why would we want this back in? Sth. Probably went wrong with your branching strategy. Please always check your diffs before requesting review.

Comment (MaxBlesch) [reply ruspy/estimation/estimation.py]: Rebased on master.

Comment (MaxBlesch) [reply ruspy/estimation/estimation.py]: Again learned something :)

File: ruspy/simulation/simulation.py [hunk 2]
```diff
@@ -51,51 +52,25 @@ def simulate(init_dict, ev_known=None, shock=None):
     beta = init_dict["beta"]
     num_periods = init_dict["periods"]
     params = np.array(init_dict["params"])
-    if "real_trans" in init_dict.keys():
-        real_trans = np.array(init_dict["real_trans"])
-    else:
-        real_trans = np.array(init_dict["known_trans"])
     if init_dict["maint_func"] == "linear":
```
Comment (peisenha) [inline ruspy/simulation/simulation.py]: ?!

File: ruspy/simulation/simulation.py [hunk 3]
```diff
@@ -105,25 +80,4 @@ def simulate(init_dict, ev_known=None, shock=None):
...
-    dist_func_shocks_maint = getattr(stats, shock[0].name)
-    dist_func_shocks_repl = getattr(stats, shock[1].name)
-    unobs[:, :, 0] = dist_func_shocks_maint.rvs(
-        **shock[0], size=[num_buses, num_periods]
-    )
-    unobs[:, :, 1] = dist_func_shocks_repl.rvs(
-        **shock[1], size=[num_buses, num_periods]
-    )
-    return unobs
+    return df, unobs, utilities
```
Comment (peisenha) [inline ruspy/simulation/simulation.py]: Can we put the `unobs` just into the simulated data frame? Then the simulated data frame is just a superset of the columns of the observed counterpart.

File: ruspy/simulation/simulation.py [hunk 4]
```diff
@@ -6,14 +6,15 @@
...
+from ruspy.simulation.simulation_auxiliary import (
+    simulate_strategy,
+    get_unobs,
+    get_increments,
+)
+from ruspy.estimation.estimation_cost_parameters import lin_cost, cost_func
 
 
-def simulate(init_dict, ev_known=None, shock=None):
+def simulate(init_dict, ev_known, real_trans_mat, shock=None):
```
Comment (peisenha) [inline ruspy/simulation/simulation.py]: There are not different concepts of transition matrix in this function, so `real` is not needed.

File: ruspy/simulation/simulation.py [hunk 5]
```diff
@@ -6,14 +6,15 @@
...
     """
     The main function to simulate a decision process in the theoretical framework of
     John Rust's 1987 paper. It reads the inputs from the initiation dictionary and
```
Comment (peisenha) [inline ruspy/simulation/simulation.py]: Docstrings are outdated, either remove at this point as 

> move fast and break things

or update.

File: ruspy/simulation/simulation.py [hunk 6]
```diff
@@ -51,51 +52,25 @@ def simulate(init_dict, ev_known=None, shock=None):
...
-            known_trans,
-            increments,
-            num_buses,
-            num_periods,
-            params,
-            beta,
-            unobs,
-            maint_func,
+    num_states = ev_known.shape[0]
+    unobs = get_unobs(shock, num_buses, num_periods)
```
Comment (peisenha) [inline ruspy/simulation/simulation.py]: I do not understand the `get_unobs` function, please upgrade it with better structure and comments.

File: ruspy/simulation/simulation_auxiliary.py [hunk 7]
```diff
@@ -213,22 +50,56 @@ def simulate_strategy_loop_known(
...
-                decision = 1
-                utility = -costs[0, 0] - costs[0, 1] + unobs[bus, period, 1]
-                new_state = increments[bus, period]
-
-            decisions[bus, period] = decision
-            utilities[bus, period] = utility
-            if period < num_periods - 1:
-                states[bus, period + 1] = new_state
+        old_state = states[bus, period]
+        if (-costs[old_state, 0] + unobs[bus, period, 0] + beta * ev[old_state]) > (
```
Comment (peisenha) [inline ruspy/simulation/simulation_auxiliary.py]: Please create temporary variables here first that have good names such as `value_replace` and `value_maintenance` for example.

File: ruspy/simulation/simulation_auxiliary.py [hunk 8]
```diff
@@ -213,22 +50,56 @@ def simulate_strategy_loop_known(
...
+        decisions[bus, period] = decision
+        utilities[bus, period] = utility
+        if period < num_periods - 1:
+            states[bus, period + 1] = new_state
+        if new_state > num_states - 10:
+            raise ValueError("State space is too small.")
     return states, decisions, utilities
+
+
+def get_unobs(shock, num_buses, num_periods):
```
Comment (peisenha) [inline ruspy/simulation/simulation_auxiliary.py]: See earlier comment on structure. This just looks way too complicated for what I suspect it is doing.

File: ruspy/test/regression_test.py [hunk 9]
```diff
@@ -26,19 +26,21 @@ def inputs():
...
     init_dict = random_init(inputs)
-    df, unobs, utilities, num_states = simulate(init_dict["simulation"])
     num_buses = init_dict["simulation"]["buses"]
     num_periods = init_dict["simulation"]["periods"]
     beta = init_dict["simulation"]["beta"]
     params = np.array(init_dict["simulation"]["params"])
     probs = np.array(init_dict["simulation"]["known_trans"])
-    v_disc_ = np.array([0.0, 0.0])
-    v_disc = discount_utility(v_disc_, num_periods, utilities, beta)
+    num_states = 200
```
Comment (peisenha) [inline ruspy/test/regression_test.py]: Why the hard-coded number?

File: ruspy/test/regression_test.py [hunk 10]
```diff
@@ -26,19 +26,21 @@ def inputs():
...
-    for bus in range(num_buses):
-        un_ob_av += unobs[bus, 0, 0]
-    un_ob_av = un_ob_av / num_buses
-    assert_allclose(v_disc[1] / (v_calc[0] + un_ob_av), 1, rtol=1e-02)
+    ev = calc_fixp(num_states, trans_mat, costs, beta)
+
+    df, unobs, utilities = simulate(init_dict["simulation"], ev, trans_mat)
+    ev_calc = calc_ev_0(ev, unobs, num_buses)
+
+    v_disc_ = np.array([0.0, 0.0])
```
Comment (peisenha) [inline ruspy/test/regression_test.py]: Why is this initialization necessary?

File: ruspy/simulation/simulation.py (hunk 2 above)
Comment (MaxBlesch) [reply ruspy/simulation/simulation.py]: As I now require the transition matrix, there is no need for specifying any probabilities in the init dict.

File: ruspy/simulation/simulation.py (hunk 3 above)
Comment (MaxBlesch) [reply ruspy/simulation/simulation.py]: But then I can only see the unobservables of the chosen alternative. This way it allows to explore afterwards decisions for each period, as I have all information on it.

File: ruspy/simulation/simulation.py (hunk 2 above)
Comment (MaxBlesch) [reply ruspy/simulation/simulation.py]: As you probably ment the maintenance function, I excuse the misunderstanding :)

File: ruspy/test/test_regression.py [hunk 11]
```diff
@@ -26,19 +26,20 @@ def inputs():
...
     init_dict = random_init(inputs)
-    df, unobs, utilities, num_states = simulate(init_dict["simulation"])
     num_buses = init_dict["simulation"]["buses"]
     num_periods = init_dict["simulation"]["periods"]
     beta = init_dict["simulation"]["beta"]
     params = np.array(init_dict["simulation"]["params"])
     probs = np.array(init_dict["simulation"]["known_trans"])
-    v_disc_ = np.array([0.0, 0.0])
-    v_disc = discount_utility(v_disc_, num_periods, utilities, beta)
+    num_states = 200
```
Comment (peisenha) [inline ruspy/test/test_regression.py]: Please integrate to `random_dict`.

File: ruspy/test/regression_test.py (hunk 9 above)
Comment (peisenha) [reply ruspy/test/regression_test.py]: Please move to random init.

File: ruspy/simulation/simulation.py (hunk 3 above)
Comment (peisenha) [reply ruspy/simulation/simulation.py]: Please add two columns.

**Pull request merged**